### PR TITLE
[#3196] i18n for Flask endpoints

### PR DESCRIPTION
**Pull request by amercader**
This is based off #3212, so you might want to review this first.

This PR includes changes to allow Flask endpoints to translate strings via [Flask-Babel](https://github.com/python-babel/flask-babel). It adds the Flask-babel requirement, hooks its middleware into the Flask stack and provides a wrapper in `ckan.common` for `_` and `ungettext`, the functions used for translations. There are also the usual cleanups and consolidations.

~~**Note:** This is marked as WIP because it relies on unpublished changes in Flask-babel regarding the support for different gettext domains: https://github.com/python-babel/flask-babel/pull/91 (cc @TkTech)~~

### CODE REVIEW

**Comment by TkTech** [comment]
Upstream changes are in, flask-babel release will be on Saturday.

File: ckan/config/middleware/flask_app.py [hunk 1]
```diff
@@ -140,6 +159,9 @@ def hello_world_post():
     for plugin in PluginImplementations(IMiddleware):
         app = plugin.make_middleware(app, config)
 
+    for plugin in PluginImplementations(IMiddleware):
```
Comment (jrods) [inline ckan/config/middleware/flask_app.py]: How come this loop is done twice?

Comment (amercader) [inline ckan/config/middleware/flask_app.py]: Thanks @jrods that was probably just a bad merge, fixed now

**Comment by amercader** [comment]
@TkTech This covers the whole i18n logic for Flask/Pylons, including support for the latest Flask-babel version.

Would you mind reviewing this first (and merging if happy)? I fear that #3239 will become too unwieldly if I keep pushing stuff to it. Thanks.

**Comment by amercader** [comment]
Now that 2.7 is branched out let's really push to get the last main Flask PRs merged into master.

This one can go first as it is small(ish) and hopefully uncontroversial. Covers changes needed to use flask-babel.

@wardi @smotornyuk if @TkTech doesn't find the time would you mind having a look and reviewing it?

Cheers

**Comment by amercader** [comment]
Bump

**Comment by smotornyuk** [comment]
Changes are pretty straightforward and i didn't noticed any problem testing it. +1 for merging it

**Comment by tino097** [comment]
good job :)

File: ckan/config/middleware/__init__.py [hunk 2]
```diff
@@ -96,10 +100,58 @@ def ask_around(self, environ):
...
+            * CKAN_LANG is the language code eg en, fr
+            * CKAN_LANG_IS_DEFAULT is set to True or False
+            * CKAN_CURRENT_URL is set to the current application url
+        '''
+
+        # We only update once for a request so we can keep
+        # the language and original url which helps with 404 pages etc
+        if 'CKAN_LANG' not in environ:
+            path_parts = environ['PATH_INFO'].split('/')
+            if len(path_parts) > 1 and path_parts[1] in self.locale_list:
```
Comment (wardi) [inline ckan/config/middleware/__init__.py]: this ignores the `root_path` setting http://docs.ckan.org/en/latest/maintaining/configuration.html?highlight=root_path#ckan-root-path

Comment (wardi) [inline ckan/config/middleware/__init__.py]: but I see this code was only moved from below..

**Comment by amercader** [comment]
Woo :tada: onwards and upwards! Next up is https://github.com/ckan/ckan/pull/3228#issuecomment-311316317 :)

**Pull request merged**